[lldb] Fix expressions that involve nested structs/classes/unions.

clayborg opened the pull request:
The LLDB expression parser relies on using the external AST source support in LLDB. This allows us to find a class at the root namespace level, but it wouldn't allow us to find nested classes all of the time. When LLDB finds a class via this mechanism, it would be able to complete this class when needed, but during completion, we wouldn't populate nested types within this class which would prevent us from finding contained types when needed as clang would expect them to be present if a class was completed. When we parse a type for a class, struct or union, we make a forward declaration to the class which can be completed. Now when the class is completed, we also add any contained types to the class' declaration context which now allows these types to be found. If we have a struct that contains a struct, we will add the forward declaration of the contained structure which can be c ompleted later. Having this forward declaration makes it possible for LLDB to find everything it needs now.

This should fix an existing issue: https://github.com/llvm/llvm-project/issues/53904

Previously, contained types could be parsed by accident and allow expression to complete successfully. Other times we would have to run an expression multiple times because our old type lookup from our expressions would cau se a type to be parsed, but not used in the current expression, but this would have parsed a type into the containing decl context and the expression might succeed if it is run again.

clayborg committed:
Fix expressions that involve nested structs/classes/unions.

The LLDB expression parser relies on using the external AST source support in LLDB. This allows us to find a class at the root namespace level, but it wouldn't allow us to find nested classes all of the time. When LLDB finds a class via this mechanism, it would be able to complete this class when needed, but during completion, we wouldn't populate nested types within this class which would prevent us from finding contained types when needed as clang would expect them to be present if a class was completed. When we parse a type for a class, struct or union, we make a forward declaration to the class which can be completed. Now when the class is completed, we also add any contained types to the class' declaration context which now allows these types to be found. If we have a struct that contains a struct, we will add the forward declaration of the contained structure which can be c
ompleted later. Having this forward declaration makes it possible for LLDB to find everything it needs now.

This should fix an existing issue: https://github.com/llvm/llvm-project/issues/53904

Previously, contained types could be parsed by accident and allow expression to complete successfully. Other times we would have to run an expression multiple times because our old type lookup from our expressions would cau
se a type to be parsed, but not used in the current expression, but this would have parsed a type into the containing decl context and the expression might succeed if it is run again.

clayborg commented:
This patch was in response to this issue:

https://github.com/llvm/llvm-project/pull/74786#issuecomment-1874921872

And is related to https://github.com/llvm/llvm-project/pull/66879, but does things a bit differently.

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -2260,7 +2269,7 @@ bool DWARFASTParserClang::CompleteTypeFromDWARF(const DWARFDIE &die,
   // Disable external storage for this type so we don't get anymore
   // clang::ExternalASTSource queries for this type.
-  m_ast.SetHasExternalStorage(clang_type.GetOpaqueQualType(), false);
+  //m_ast.SetHasExternalStorage(clang_type.GetOpaqueQualType(), false);
```
This should not be in here and I will revert this. I was testing to see if I removed this, if I would get a `find 'B' in 'A'` call to the `ClangASTSource::FindExternalVisibleDecls(...)`, but it never happened. I was hoping we could just leave the external AST source enabled for classes and that would allow us to find the contained types efficiently, so that was my original approach.

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -3106,10 +3115,39 @@ void DWARFASTParserClang::ParseSingleMember(
+      break;
+  }
+  return false;
+}
```
I tried to find this function, but was able to find it because `llvm::dwarf::isType` uses the #include style to make the contents which is not readable or grep'able:
```
inline bool isType(Tag T) {
  switch (T) {
  default:
    return false;
#define HANDLE_DW_TAG(ID, NAME, VERSION, VENDOR, KIND)                         \
  case DW_TAG_##NAME:                                                          \
    return (KIND == DW_KIND_TYPE);
#include "llvm/BinaryFormat/Dwarf.def"
  }
}
```
No wonder I couldn't find it with a code grep! Thanks for the pointer, I will use this

Michael137 replied:
(kudos to @felipepiovezan who coincidentally told me about it for something else today morning)

clayborg committed:
Fix review comments.

Fixes:
- revert the code that commetned out turning off external AST for a class, this was from initial testing
- remove IsTypeTag() and use llvm::dwarf::isType()

clayborg committed:
Run clang-format

Michael137 reviewed (approved):
LGTM

dwblaikie commented:
Thanks for the fix! I did test this patch with both our regression issue, and also allowing us to remove the double-lookup working around #53904 and can confirm it addressed both issues. Thanks again!

clayborg commented:
> Thanks for the fix! I did test this patch with both our regression issue, and also allowing us to remove the double-lookup working around #53904 and can confirm it addressed both issues. Thanks again!

Glad it worked! It will also be a huge improvement in the expression parser and avoid these weird expression parser bugs where it doesn't work sometimes and does other times, so it was totally worth fixing. I am also glad to know that the type lookups are being so much more efficient. Having a repro case is always the best way to help us fix bugs, so thanks for posting the details so we can repro easily.

DavidSpickett commented:
One of the new tests fails on Windows:
```
FAIL: test_static_in_nested_structs_dwarf (TestNestedExpressions.NestedExpressions)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\packages\Python\lldbsuite\test\lldbtest.py", line 1696, in test_method

    return attrvalue(self)

           ^^^^^^^^^^^^^^^

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\API\commands\expression\nested\TestNestedExpressions.py", line 44, in test_static_in_nested_structs

    self.expect_expr("A::B::C::enum_static",

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\packages\Python\lldbsuite\test\lldbtest.py", line 2508, in expect_expr

    value_check.check_value(self, eval_result, str(eval_result))

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\packages\Python\lldbsuite\test\lldbtest.py", line 300, in check_value

    test_base.assertSuccess(val.GetError())

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\packages\Python\lldbsuite\test\lldbtest.py", line 2543, in assertSuccess

    self.fail(self._formatMessage(msg, "'{}' is not success".format(error)))

AssertionError: 'error: Couldn't look up symbols:

  A::B::C::enum_static

' is not success
```
But given the other issues we've had lately, it's not surprising.

I've skipped it for now - https://github.com/llvm/llvm-project/commit/ba4cf31facdaf9bb9943c057d325ff0968331e9a.

dwblaikie commented:
> > Thanks for the fix! I did test this patch with both our regression issue, and also allowing us to remove the double-lookup working around #53904 and can confirm it addressed both issues. Thanks again!
> 
> Glad it worked! It will also be a huge improvement in the expression parser and avoid these weird expression parser bugs where it doesn't work sometimes and does other times, so it was totally worth fixing. I am also glad to know that the type lookups are being so much more efficient. Having a repro case is always the best way to help us fix bugs, so thanks for posting the details so we can repro easily.

Turns out it actually got worse, see https://github.com/llvm/llvm-project/issues/79668 - now whether or not a given nested type is accessible via name lookup depends on which copy of the outer type is loaded first & there's no workaround that I know of - the lookup then fails/succeeds permanently.

Be good to get this addressed - as now it's not possible to lookup these nested members at all reliably (even the "repeat the lookup" workaround doesn't work - maybe there are other workarounds? But none that I can think of)

clayborg commented:
The main issue we have is they way that clang works currently with the external AST source. We currently create a forward declaration type "struct Foo;" and we allow it to be completed once. Once it is completed clang expects the struct type to have everything that it needs, like all contained types (forward decls are allowed), otherwise we will never find that type as the external AST source thinks it is done. 

I would like to avoid having to find _all_ copies of a struct type and parse all contained types in all of matching entries from everywhere in the DWARF if possible. If we did this then we would be parsing tons of debug info to add any and all potentially contained types and that would cause way too much debug info to be parsed.

The best solution I can think of is to have the compile continue to call the external AST source to say "please find this type within this struct/class". These calls don't happen today, but if they did, we could do this very efficiently without having to populate all contained types into every struct/class anytime we complete the type.

So right now the flow is:
- any struct/class gets created as a forward declaration and has the external AST source enabled
- the external AST calls to complete the type when needed by expression parser or LLDB APIs
- we complete the type
- we remove the external AST flag on the type so no more external AST calls happen on this class

The new flow could be:
- any struct/class gets created as a forward declaration and has the external AST source enabled
- the external AST calls to complete the type when needed by expression parser or LLDB APIs
- we complete the type
- we leave the external AST enabled so we can get called for contained types

The last item here would require modifications to clang which could be debugger specific changes that are only enabled in a special debug compiler mode so they don't affect the actual compiler. Thoughts?

dwblaikie commented:
How'd this work before your recent changes, then - when each repeated query would get one level further down in the nesting? How'd that work given the clang limitations you're describing?

In any case, the extra clang requirements here seem like they might be of uncertain cost/maintainability (if it's only updating one place that everyone calls through - great, but if it's updating multiple callers, and the risk that new callers miss this handling - that seems like a maintainability problem) which is worrying. So, I'd be uncertain about that direction without more info - and with that info, if it is just one codepath, yeah, maybe it's quick enough to address the regression.

But if it'd require substantial reengineering to clang to get this working - this seems an unfortunate regression to carry in the interim, and it might be worth reverting to the previous (differently buggy, but at least work-around-able) behavior.

And if this would require updating many callers in clang (unless such updates include an API change that would make it difficult to accidentally introduce a new caller that didn't handle things correctly for lldb), I'd worry about the future stability of such a change & might be inclined towards the less efficient "search all the DWARF" thing.

clayborg commented:
> How'd this work before your recent changes, then - when each repeated query would get one level further down in the nesting? How'd that work given the clang limitations you're describing?

It was because we were actually parsing and converting many extra types. Anytime we did a type query for a given type, we would end up converting all of those types into the clang AST. So for `A::B::C`, the first time we would run this we would actually parse and convert `A` into the AST, but this would cause a bogus type query to then get sent looking for `B` at the translation unit level. But since our old type parsing code always created the types whose basename matched without looking at wether the decl context matches, it would end up parsing and converted all types whose basename matched `B` into the AST. When we incorrectly parsed the `B` type, we would and up actually putting it into `A` as the decl context, so the next time we ran an expression, `B` would already be in the definition for `A` and the expression would then work. So it was due to us parsing too many types, even if they were incorrectly scoped, but we would always create them correctly inside of the correct decl context...

> In any case, the extra clang requirements here seem like they might be of uncertain cost/maintainability (if it's only updating one place that everyone calls through - great, but if it's updating multiple callers, and the risk that new callers miss this handling - that seems like a maintainability problem) which is worrying. So, I'd be uncertain about that direction without more info - and with that info, if it is just one codepath, yeah, maybe it's quick enough to address the regression.

Yeah, it would be great if anyone with compiler experience could check and see if this is controllable and easy to fix. I was thinking we could add new APIs to the External AST that are marked as "only override these if you are a debugger"
> 
> But if it'd require substantial reengineering to clang to get this working - this seems an unfortunate regression to carry in the interim, and it might be worth reverting to the previous (differently buggy, but at least work-around-able) behavior.

I would rather have the current problem over the previous one. Are we mostly worried about template classes here? Or do all classes, even non templated ones, have this same issue?

> And if this would require updating many callers in clang (unless such updates include an API change that would make it difficult to accidentally introduce a new caller that didn't handle things correctly for lldb), I'd worry about the future stability of such a change & might be inclined towards the less efficient "search all the DWARF" thing.

I mean, in the expression parser we clearly say with `A::B`: "I am looking for something named `B` within the context of `A`" and only in the classes, structs unions or enums. Without this fix, we will end up making a global lookup for `B`, which might be ok for some things, but imagine typing `A::iterator`. If we go back to the previous bug, we will do a lookup for `A` at the translation unit level and when the auto lookup for `iterator` fails to be found within `A`, a type search for `iterator` at the translation unit level will take place.

Michael137 commented:
> But if it'd require substantial reengineering to clang to get this working - this seems an unfortunate regression to carry in the interim, and it might be worth reverting to the previous (differently buggy, but at least work-around-able) behavior.

To clarify, the regression @dwblaikie reported in https://github.com/llvm/llvm-project/issues/79668 was introduced in https://github.com/llvm/llvm-project/pull/74786, not in this current PR. IIUC, this PR just *partially* fixed https://github.com/llvm/llvm-project/issues/53904?

> The new flow could be:
> 
> * any struct/class gets created as a forward declaration and has the external AST source enabled
> * the external AST calls to complete the type when needed by expression parser or LLDB APIs
> * we complete the type
> * we leave the external AST enabled so we can get called for contained types
> 
> The last item here would require modifications to clang which could be debugger specific changes that are only enabled in a special debug compiler mode so they don't affect the actual compiler. Thoughts?

Following overview of the clang<->lldb call-chain might be useful: https://github.com/llvm/llvm-project/issues/53904#issuecomment-1877114034. Not sure I fully understand the "we leave the external AST enabled so we can get called for contained types" idea. Happy to chat about how to make this work though

clayborg commented:
> > But if it'd require substantial reengineering to clang to get this working - this seems an unfortunate regression to carry in the interim, and it might be worth reverting to the previous (differently buggy, but at least work-around-able) behavior.
> 
> To clarify, the regression @dwblaikie reported in #79668 was introduced in #74786, not in this current PR. IIUC, this PR just _partially_ fixed #53904?
> 
> > The new flow could be:
> > 
> > * any struct/class gets created as a forward declaration and has the external AST source enabled
> > * the external AST calls to complete the type when needed by expression parser or LLDB APIs
> > * we complete the type
> > * we leave the external AST enabled so we can get called for contained types
> > 
> > The last item here would require modifications to clang which could be debugger specific changes that are only enabled in a special debug compiler mode so they don't affect the actual compiler. Thoughts?
> 
> Following overview of the clang<->lldb call-chain might be useful: [#53904 (comment)](https://github.com/llvm/llvm-project/issues/53904#issuecomment-1877114034). Not sure I fully understand the "we leave the external AST enabled so we can get called for contained types" idea. Happy to chat about how to make this work though

When we parse a class, we enable the external AST support on the CXXRecordDecl object. When we are asked to complete the type it will eventually make it down into `DWARFASTParserClang::CompleteTypeFromDWARF(...)` which will call:
```
  // Disable external storage for this type so we don't get anymore
  // clang::ExternalASTSource queries for this type.
  m_ast.SetHasExternalStorage(clang_type.GetOpaqueQualType(), false);
```
So we never get a query again to find anything inside of this class. That being said, I am not sure anything ever was doing another external AST lookup in this class after we completed it. But if we are able to add debugger specific support into the external AST virtual interface, we might be able to gets asked "withing this 'class A' please find me a contained type" and then modify the compiler to use the external AST source when a lookup in a contents fails. 

We currently modified it so that when we parse a struct/union/class type, we also parse all contained child DIE types so that they are available. This is failing when some classes have contained types defined and others don't because we only add the contained types for the struct/class/union DIE we used to create the type. If we can modify the compiler to make an external lookup call only when debugging, we can avoid this need to parse all contained types and wait until the expresion parser asks for that type.

Michael137 commented:
>  That being said, I am not sure anything ever was doing another external AST lookup in this class after we completed it. But if we are able to add debugger specific support into the external AST virtual interface, we might be able to gets asked "withing this 'class A' please find me a contained type" and then modify the compiler to use the external AST source when a lookup in a class decl context fails.

Ah I see, it sounds like what you want is a call to `ExternalASTSource::FindExternalVisibleDeclsByName` when [this lookup](https://github.com/llvm/llvm-project/blob/c6b5ea339d9f257b64f4ca468e447f0e29a909a4/clang/lib/Sema/SemaLookup.cpp#L1123-L1130) into a `DeclContext` fails:
```
static bool LookupDirect(Sema &S, LookupResult &R, const DeclContext *DC) {
  ...
  DeclContext::lookup_result DR = DC->lookup(R.getLookupName());

  <<<< Ask LLDB to look for possibly contained types >>>>

  for (NamedDecl *D : DR) {
    if ((D = R.getAcceptableDecl(D))) {
      R.addDecl(D);
      Found = true;
    }
  }
```

dwblaikie commented:
Be great to see D101950 picked up again, so I'm glad to hear that's being looked at.

Be nice to get some way to address this regression sooner rather than later... 

& yeah, I'd argue that classes derived from DWARF are not complete - nested types, member function templates, special member functions, and... static members I think - none of those things are consistently emitted in every copy of a type. So we probably shouldn't turn off the queries back into lldb for a type ever, really.
